File: sol/solution/0001-0100/0008. String to Integer (atoi)/Solution.c
int myAtoi(char *s)
{
    int i = 0, sign = 1, result = 0;
    while (s[i] == ' ')
        i++;
    if (s[i] == '+' || s[i] == '-')
        sign = (s[i++] == '-') ? -1 : 1;
    while (s[i] >= '0' && s[i] <= '9')
    {
        int digit = s[i++] - '0';
        if (result > (INT_MAX - digit) / 10)
            return sign == 1 ? INT_MAX : INT_MIN;
        result = result * 10 + digit;
    }
    return result * sign;
}

File: sol/solution/0001-0100/0008. String to Integer (atoi)/Solution.ex
defmodule Solution do
  @int_max 2_147_483_647
  @int_min -2_147_483_648
  def my_atoi(s) do
    s = String.trim_leading(s)
    {sign, rest} = case s do
      "-" <> tail -> {-1, tail}
      "+" <> tail -> {1, tail}
      _ -> {1, s}
    end
    digits = take_digits(rest, "")
    if digits == "", do: 0, else: parse_int(digits, sign)
  end
  defp take_digits(<<c, rest::binary>>, acc) when c in ?0..?9 do
    take_digits(rest, acc <> <<c>>)
  end
  defp take_digits(_, acc), do: acc
  defp parse_int(digits, sign) do
    {int, _} = Integer.parse(digits)
    result = sign * int
    cond do
      result < @int_min -> @int_min
      result > @int_max -> @int_max
      true -> result
    end
  end
end